### DISCUSSION TITLE
Last Push to V4 Release

### DISCUSSION THREAD

**Discussion by jeremydmiller** [JasperFx/marten#1734] [General]
Just to have an open forum to talk through what's left to do for V4, and what we might cut from scope to V4.1 so we can get this thing out the door. I (Jeremy) am writing the first shot at this just to facilitate conversation. With as long as V4 development has gone on, how long it was delayed starting in the first place, and me running out of steam a little bit, I think we're approaching the time to triage remaining scope and only play things that are either breaking changes or really important like the new website. I also think we (me) need to leave some energy to be writing blog posts and sample applications at the end of this to promote V4 to make all the work worth it.

The [open issues left for V4 are here](https://github.com/JasperFx/marten/milestone/28).

After going through the issues, here's how I think I'd break things down:

## Documentation

I think this is the biggest area of remaining effort, and probably what's most important

* **Definitely in** -- the move to using [VitePress](https://vitepress.vuejs.org/) for the documentation and getting a new, spiffier layout and styling. I'd say this is #1401 
* **In**: #1395, #1396, #1398
* #521 will just be the new documentation on the V4 `ViewProjection`
* #1555 is just a reminder of things that need to be doc'd
* 

### Discuss

* #1435 -- what does this really mean? could fall to >4.0
* #1611 -- seems important, folks ask for it, might require breaking changes
* #1613 -- need more investigation. Jeremy thinks this could be added later without breaking changes
* #1615 -- I think this is more about documentation than anything new, but we need to explore it anyway
* #1665 -- Not huge, kinda need to do for backward compatibility
* #1671 -- eliminate the aggregate type definition in `StartStream<T>()`? Doesn't do anything today. Just a marker that isn't paid attention to otherwise. **Breaking change**
* #1677 -- I want this personally, but I wanna make some Oakton changes first
* #1689 -- I'm not 100% sure what there is to do, or if we should. Smells like edge cases users could do w/ custom projections
* #1717 -- sharding the projection async work. I think this would be hugely valuable, but a bridge too far
* 

### Do for V4

* Event Metadata should be in. #780. The only thing I'd add here is to say I'd still like to a do a sample app before v4 off to the side with Jasper + Marten that supports the open telemetry spec from HTTP endpoints & messaging endpoints with Marten activity being logged as part of that. I think that would help prove out the logging and metadata for Marten v4
* #770, partitioning support for the event & stream tables
* #1612 -- Archiving event streams. Seems relatively small the way Oskar, Babu, and Jeremy discussed doing it. Big advantage for scalability
* #1647 -- Lamar bug
* #1679 -- Lamar bug
* #1718 -- at least for the hot/cold failover mode. It's doable w/ advisory locks. Gets us to where folks can incorporate the async daemon better than they do today in app clusters
* #1723 -- might be a bug, might not be an issue. Just needs investigation
* #1731 -- we don't have this in V3, but it seems important

### Push off to later

* #1230. If this is just a copy now of EventStoreDb's RSS subscription, I'm not enthusiastic about taking this on in Marten core
* #1670 -- I vote later
* #1680 & #1681 -- nice to do
* #1682 -- I think this is more work than it's worth
* #1714 -- my vote is to push this off. Really to be able to use build products in one projection in another projection
* #1720 & #1717 -- Definitely going to come later. I think I'm back to wanting to do this inside of Jasper rather than just Marten. And it's **big**

## Extracting the Database Helpers to its Own Library

This is #1467. To be honest, I want to do this because I want to reuse a lot of the code here in Jasper, and it is generic enough to move out of Marten itself. I'm thinking of pulling out `CommandBuilder`, the `IFeatureSchema` types, and utility extension methods on Npgsql.

## Logging

* Let's do #1686 as a quick win

## Multi-Tenancy

#1451 & #1460 are both relatively easy. I've just been putting them off thinking we could get the folks who reported them to help

### To discuss

* #1539 
* #1230 -- if we do this, we need to put a read only interface over the top of `StoreOptions`
* #585 -- *I'd really like to see this in V4, but it's an additive change*
* #1391 -- is there any way to "close" this, or do we convert that to an ongoing discussion?
* #1390 -- same as the above issue
* #1397 -- seems doable
* #1283 -- there's demand from our F# users, but it's an additive change that could come later
* #1564 -- would be breaking changes

### In for V4

* #1470, the big sweep to move any possible type into `internal` we can to reduce the public API space
* #1570 -- improvement to patching
* 

### Later
* #1262. I don't honestly understand this one yet
* #1315 -- it's a limitation of the serialization, not Marten per se. I'm not sure we can fix this
* #1618 & #1647 -- Linq helpers for raw event selectors. I vote this goes later
* #1660 -- I think it would be cool, but it's not a breaking change, so later?
* #1661 -- I vote later. Could be a way to optimize the async projections too
*

**Comment by oskardudycz** [comment]
@jeremydmiller - 
Documentation

For me, the most important is to deliver new docs together with the version switcher. If it contains the features that we had, then from my perspective we can polish it later. Based on what @mysticmind wash showing, it'll be still a vast improvement and "wow effect" for marketing. Pages with links to blog posts are nice to have for me. We should have it asap, but that's not a blocker for me.

I'd be happy to work on the Event Sourcing docs. We should go with our user a journey from the basic concepts like events, streams, and appending and aggregation. Then explain snapshots (inline projections) and projections. I planned to do it earlier, but work on v4 stopped me as it could be made obsolete. Of course, @jeremydmiller, if you prefer to do it on your own, I won't interfere with that.

Event Store
Discuss

- #1435 -- I think we need to cover two main scenarios - migration of the data from other event store or relational/document data and integration with queuing systems. Probably we should have a middleware that would allow the user to do some data transformations. For the event store, we should provide either the custom version or aggregate by different criteria. This is especially important for queues integration, as they may not maintain order. So allow the user to insert raw event data without going through all checks and logic. 
- #1611 - could you elaborate what's projection snapshot? I don't know this request and this concept
- #1613 - yes, this can be added in v4.1 this shouldn't be a breaking change, as it's adding a middleware/pipeline to the deserialisation part. You can check my recent discussion on this topic https://discuss.eventstore.com/t/deserialisation-in-c/3091
- #1615 - Agree, we should provide documentation and samples, and it should be enough
- #1665 - I think that this should be a part of V4 - @ericgreenmix found that for the high load without that we may have concurrency issues
- #1671 - I don't think that we should remove it. We could make it opt-in, but I think we should keep it. What's more, if we allow having AggregateType on the events table, together with partitioning (#770), it'll add the possibility to have events partitioned by aggregate type. The queue integration/outbox pattern will open case, e.g. to publish aggregate events to the same Kafka topic.
- #1677 - Btw. do we know how many people are using Marten CLI? TBT I've never used it.
- #1689 - I was using that in my project to make sure that I won't be (re)creating projection on some race conditions or connection with deleting events. Especially with integration with events from other modules by message bus, it's useful. I agree that's related to edge cases, but it can save you some time and data integrity
- #1717 - I think that's a nice to have for v4

Do for V4

- #780 - Yes, that's a must-have. This is the most significant from the features left in ToDo. @mysticmind do you still plan to work on that?
- #770 - I think we can deliver that in a non-breaking manner as opt-in potentially in v4.1. I'd be happy to work on that.
- #1612 - Yes, together with #770, this would be nice to have. I think that this can also be part of v4.1
- #1647, #1679 -- Looks like both are fixed - nice @jeremydmiller
- #1718 - That looks quite important, however doing proper tests may take a lot of time, I'm okay with moving it to v4.1	
- #1723 - yup, we need to understand that at least
- #1731 - It's essential, but I think that we could deliver that later in v4.1 in a non-breaking manner

Push off to later

- #1230 - @jeremydmiller are you sure that's the right issue number?
- #1670 - yes, we can add it later
- #1680 & #1681 - nice to do - I agree
- #1682 - What is "self-aggregators"?
- #1714 - yes, this is an edge case that probably should be solved on the application layer (usually).
- #1720 & #1717 - For me, sharding is not so necessary. Zero downtime - if I understand that correctly is a must-have eventually. However, we may push it for the next release

**Comment by oskardudycz** [comment]
## Extracting the Database Helpers to its Own Library

- #1467 - I'm okay with that, but this should probably go to v4 as it'll be a breaking change

## Logging

- #1686 - I see this is done - sweet!

## Multi-Tenancy

- #1451 & #1460 - I think that we should discuss the elegant approach to the API before doing the work

## Miscellaneous
To discuss

- #1539 - this would be cool to have - @mysticmind do you think this could be included later as non-breaking opt-in? 
- #1230 - I'm okay with putting the read-only interface over the top of StoreOptions
- #585 - This would be cool from the marketing perspective. I could try to tackle that after my work on System.Text.Json. I agree that this should not block the v4 release if it's not ready
- #1391 - I think that as we're not planning to do any action on this topic, then we can close it...
- #1390 - Are you sure that you linked the right issue?
- #1397 - yes, it's doable, the question is if we want to do it, and what would be our goal for that (as we plan to close #1391)
- #1283 - I think that with new projections etc. Marten has good potential in the functional community. It's important but can go in v4.1
- #1564 - I agree, it would be good to clean up that in v4 

In for V4
- #1470 - yes, we need to make sure that we cleared the API and only stuff that should be public is public, so we could provide more changes in a non-breaking way not being blocked by "maybe someone is using it."
- #1570 - could you elaborate? I'm not following what's the case here

Later

- #1262 - it might be just a simple syntax change, as it looks breaking we may consider that for v4
- #1315 - We could store long as string
- #1618 & #1647 - Agree, nice to have
- #1660 - can go later - agreed.
- #1661 - yes, can go later

**Comment by oskardudycz** [comment]
Ah, I think that my PR with System.Text.Json should go in - it's breaking https://github.com/JasperFx/marten/pull/1685

**Comment by Hawxy** [comment]
Excited for the new docs :). Would it be worth having a section dedicated to "marten in production" scenarios? ie an example of how a CI/CD pipeline should be setup to deal with marten's schema changes over time.

#1539 - Would be great for those of us with clients that have security-mature policies and are uncertain about using a multi-tenanted single-database solution.

#770 + #1612 - Highly valuable for long-term scalability, yes please!

**Comment by jeremydmiller** [comment]
Hey all, it's time to once again play "what's it going to take to finish V4?"

I've got a relatively free day tomorrow at work, and I can throw maybe 3-4 hours of uninterrupted time at Marten. Here's what I see is left **besides the new documentation** for development:

* Event Metadata (#780). @mysticmind Do you still want to take this? I could divert and pound this out if you want. 
* #1230 - I can take this. Shouldn't be any big deal
* @oskardudycz is already in flight with #770 
* I'm still interested in doing #1467, but we shouldn't tough that until #770 is done
* #1470 is a last minute sweep of the codebase, and I don't anticipate that being a big deal
* #1283 - the F# discriminated union querying. I think this could actually be done now w/ a Marten add on, but it'd still be good to prove that out. I kinda want some F# dev help here
* #1262 -- the title might be misleading. It's really "Make Marten work with TimescaleDB". I wrote that I thought it was a breaking change because it would change the upsert functions. *I want to vote to move this to 4.1*
* #1539 -- something similar came up recently where it would be advantageous to set up SQL operations to take place upon a session being initialized
* #1315 -- @oskardudycz suggested storing longs as strings. We didn't attempt to tackle any kind of optimized patching story in V4. *I'd like to make anything around patching be a V4.1 thing and put this off*
* #1435 - I'm thinking we do something w/ binary copy to make it fast. *I'm voting to push this to post V4*
* #1612 - Archiving event streams. As we discussed on a call, I think this in combination w/ #770 is a big win for scalability. I'd be happy to take that on whenever, but I don't know if it would be better to let #770 go first
* #1611 - We haven't talked about this at all, but I think we should on our next Marten team call. I don't have any thoughts on how to do this or requirements offhand.
* #1615 - Recipe for testing event projections. I'm going to take this tomorrow, and do something similar to [Alba](https://jasperfx.github.io/alba) for projections
* #1731 - I've got some ideas around this one, and I don't think it's big. This will be in my next wave of work
* #1718 - the Async Daemon leader election using advisory locks. I don't think this is very much code at all, but it's gonna be a pain to test. I'm happy to work on this because it'll be similar to work I've done in Jasper
* #1735 - the memory usage w/ Roslyn. At the very worst, I have a workaround. Really just need to run the tests with memory usage tracing and see what's hanging on to the memory

I'm sure we'll come up with other stuff, plus the documentation & new website.

Thoughts???

**Reply by jeremydmiller** [reply]
And I completely missed #1613. I think I'd ask @oskardudycz to speak to that next time we all jump on a call

**Reply by mysticmind** [reply]
@jeremydmiller with regards to #780, I was planning to take a crack at it by early next week, besides the plan was to understand the meta data portion so that I can also add the documentation for it. If you think, you can take it up, please do, don't want to keep it as a blocker for v4.

**Reply by jeremydmiller** [reply]
@mysticmind If you could take #780 and the docs on metadata, that would be a huge contribution. There's plenty of other things for me to be doing in the meantime. Yell if you have any questions or suggestions about what's already been done on the metadata.

I don't think this would help timewise, but I'd *love* to switch back to Jasper soon, push through an OpenTelemetry implementation there, then drop Marten into that and prove that the new metadata support in Marten works well with OT tracing.

**Reply by mysticmind** [reply]
@jeremydmiller sure, will get it done 👍

**Comment by jeremydmiller** [comment]
Just time for an update on V4, because it's taken much, much longer than I think anyone would have hoped for. The [milestone](https://github.com/JasperFx/marten/milestone/28) is getting smaller at least.

Me, personally, I'd like to declare success on my development work by mid-April and focus on documentation updates. In a perfect world, **I'd love for us to shoot for v4 escaping into the wild by the end of April**. Commence laughing at me.

Here's how I see the remaining **development** work:

* @mysticmind is working on the Event Metadata story (#780) and I think that's close
* @Hawxy is adding some nullability checks throughout the codebase, and we'll need to do that before V4 since it changes the public API
* @JohnCampionJr is working on the F# discriminated union request #1283
* @oskardudycz is working on #1784 that's necessary for us to support event projections where one event could apply to multiple projected views
* We'll need to tackle the usage of Postgresql table partitioning in our schema management. Once that works, that opens up event store partitioning (#770) and the event store stream archiving (#1612). Somewhat related to that is moving the generic ADO.Net helper and Postgresql schema management functionality to a new library (#1467). I'm still interested in making this happen, but I think this happens right after we make our next alpha release. I may start that work in a branch soon.
* The mechanism to import events is probably dependent upon #780 and also #1612 and #770. There *is* user demand for this, but it could float to 4.1 in my opinion because it's not a breaking change. 
* We'll do the `internal` accessor hunt (#1470) and fill in Xml API comments at the last minute. I'm not too worried about that. 
* It's not on the milestone yet, but near the end let's do a find and replace on `.ConfigureAwait(false)` and wipe that out of the codebase
* I (@jeremydmiller) will take a look at the memory usage issue (#1735) with dotProfile before V4. I'm not too worried about this, and if we can't fix it, I've got a workaround in mind from Jasper to pre-build the generated source code ahead of time. 
* #1539 would be a cool feature, and might not be very difficult. 
* I think we need to have a conversation about some of the remaining event store work. We have no analysis done yet on Projection Snapshots #1611 (basically sampling projected views by persisting it every so many versions).
* We also need to talk a little more about event versioning and upcasting (#1613). I think this could be done w/o any breaking changes and might be a candidate for 4.1 if we can't squeeze it in for 4.0

**Reply by johncj-improving** [reply]
Is that a typo wrt #1283 or is someone else also working on it?

**Reply by jeremydmiller** [reply]
Typo on my part, sorry about that:)

**Reply by johncj-improving** [reply]
My fix for #1283 is completely external to the Marten codebase. There's no need to hold up the release of 4.0 for it. In fact, it is somewhat dependent on the completion of 4.0. In particular, the implementation of #780 and #1784 will have an impact on my solution.

**Comment by jeremydmiller** [comment]
## The Very, Very Last Push

At this point I want to cut bait, push every non-breaking change to 4.0, and start writing documentation updates.

I did some more research on using table partitioning against the event table, and I think it's a viable strategy. I also think it would tie in very well with some of the "async daemon sharding" stories that have already gotten pushed to the 4.1 milestone. What I propose at this point is to:

* Let @oskardudycz finish #1784 in a simple way for now
* Knock out all the documentation related issues in the 4.0 milestone next week (https://github.com/JasperFx/marten/milestone/28)
* Do the "Archiving Event Streams" story (#1612) in a simplistic way just to get the new `is_archived` column and related changes in place, but punt on table partitioning until 4.1. I've gone back and forth, but I vote to do the event store archiving with the assumption that it works with partitioning. **We should discuss this decision to punt the partitioning**
* I'll deal with the memory issue this or next week (#1735). I have a workaround strategy to pre-generate the source code ahead of time as a last ditch effort to fix this.
* Push #770, #1539 (I think this could be mostly done w/ the `ISessionFactory` thing, plus some extensions to Weasel for RLS), #1435, #1611 to 4.1.
* Move PLV8 related stuff to an add on library #1812. **This would impact the Patch() functionality!**
* New little story on projection configuration usability: #1813
* Cut a Marten 4.0.0-rc-1 **at the end of this week**
* Shoot for the new website to be ready-ish by the end of next week

The 4.1 release would end up being a lot of optimization of the event store functionality that's getting missed by 4.0. I *think* that 4.2 could be doing an optimized, non-plv8 using version of the Patch() API. 

4.0 has been in flight for over a year of coding and 2 years of planning, and it's starting to feel like Duke Nukem Forever. I think I want to move to documentation and the inevitable last minute API changes that should result from doing the documentation and sample apps.

There's an artificial time limit here. I'm starting a new job the last week of May, and I'd dearly love V4 to basically be done before I start the new job. I think I'll have more energy for OSS in my new role, but who knows?

**Reply by oskardudycz** [reply]
The plan looks reasonable 👍 I don't have points where I'd disagree 😉
